File: src/internal/api/manage.go
package api

import (
	"context"
	"errors"
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/ranaroussi/minigun/internal/models"
	"github.com/ranaroussi/minigun/internal/store"
	"github.com/ranaroussi/minigun/internal/tmpl"
	"github.com/ranaroussi/minigun/internal/token"
)

type manageContext struct {
	Token     string
	Send      *models.Send
	List      *models.List
	Contact   *models.Contact
	Sub       *models.Subscription
	Company   *models.Company
}

func (s *Server) loadManageContext(ctx context.Context, tokenStr string) (*manageContext, error) {
	t, err := token.Verify(s.cfg.HMACSecret, tokenStr)
	if err != nil {
		return nil, errors.New("Invalid or expired manage link.")
	}
	snd, err := s.store.GetSend(ctx, t.SendID)
	if err != nil {
		return nil, errors.New("Send not found.")
	}
	sub, err := s.store.GetSubscriptionByID(ctx, t.SubscriptionID)
	if err != nil {
		return nil, errors.New("Subscription not found.")
	}
	contact, err := s.store.GetContactByID(ctx, sub.ContactID)
	if err != nil {
		return nil, errors.New("Contact not found.")
	}
	list, err := s.store.GetListByID(ctx, sub.ListID)
	if err != nil {
		return nil, errors.New("List not found.")
	}
	if list.CompanyID == "" {
		return nil, errors.New("This list is not associated with a company; manage page is not available.")
	}
	company, err := s.store.GetCompanyByID(ctx, list.CompanyID)
	if err != nil {
		return nil, errors.New("Company not found.")
	}
	return &manageContext{
		Token:   tokenStr,
		Send:    snd,
		List:    list,
		Contact: contact,
		Sub:     sub,
		Company: company,
	}, nil
}

func (s *Server) handleManageGet(w http.ResponseWriter, r *http.Request) {
	tokenStr := chi.URLParam(r, "token")
	mc, err := s.loadManageContext(r.Context(), tokenStr)
	if err != nil {
		s.renderManagePage(w, tmpl.ManageData{Error: err.Error()})
		return
	}
	states, err := s.store.GetCompanyListsWithSubscription(r.Context(), mc.Company.ID, mc.Contact.ID)
	if err != nil {
		s.renderManagePage(w, tmpl.ManageData{Error: "Failed to load preferences."})
		return
	}
	s.renderManagePage(w, tmpl.ManageData{
		Token:       tokenStr,
		Email:       mc.Contact.Email,
		CompanyName: mc.Company.Name,
		Lists:       states,
	})
}

func (s *Server) handleManagePost(w http.ResponseWriter, r *http.Request) {
	tokenStr := chi.URLParam(r, "token")
	if err := r.ParseForm(); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	mc, err := s.loadManageContext(r.Context(), tokenStr)
	if err != nil {
		s.renderManagePage(w, tmpl.ManageData{Error: err.Error()})
		return
	}

	companyLists, err := s.store.ListsForCompany(r.Context(), mc.Company.ID)
	if err != nil {
		s.renderManagePage(w, tmpl.ManageData{Error: "Failed to load preferences."})
		return
	}

	checked := map[string]struct{}{}
	for _, v := range r.PostForm["list"] {
		checked[v] = struct{}{}
	}

	desired := make([]store.SubscriptionChange, 0, len(companyLists))
	for _, l := range companyLists {
		_, ok := checked[l.ID]
		desired = append(desired, store.SubscriptionChange{
			ListID:     l.ID,
			Subscribed: ok,
		})
	}

	deltas, err := s.store.ApplySubscriptionChanges(r.Context(), mc.Contact.ID, desired)
	if err != nil {
		s.log.Error("apply subscription changes", "err", err)
		s.renderManagePage(w, tmpl.ManageData{Error: "Failed to save preferences."})
		return
	}

	for _, d := range deltas {
		if d.WasSubbed && !d.NowSubbed {
			sub, err := s.store.GetSubscription(r.Context(), d.ListID, mc.Contact.ID)
			if err != nil {
				continue
			}
			if _, err := s.store.RecordUnsubscribeEvent(r.Context(), &mc.Send.ID, sub, mc.Contact.Email); err != nil {
				s.log.Error("record unsub event from /manage", "err", err)
			}
		}
	}

	s.renderManagePage(w, tmpl.ManageData{
		Done:        true,
		Email:       mc.Contact.Email,
		CompanyName: mc.Company.Name,
		Deltas:      deltas,
	})
}

func (s *Server) renderManagePage(w http.ResponseWriter, data tmpl.ManageData) {
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	if data.Error != "" {
		w.WriteHeader(http.StatusBadRequest)
	}
	if err := tmpl.Manage.Execute(w, data); err != nil {
		s.log.Error("render manage page", "err", err)
	}
}

File: worker/src/routes/manage.ts
import { Hono } from 'hono';
import { Env } from '../env';
import { InvalidTokenError, verify } from '../lib/token';
import { ManagePage } from '../pages/manage';
import { getContactByID } from '../store/contacts';
import { getCompanyByID, listsForCompany } from '../store/companies';
import { getListByID } from '../store/lists';
import {
  applySubscriptionChanges,
  getCompanyListsWithSubscription,
} from '../store/manage';
import { getSend } from '../store/sends';
import { getSubscription, getSubscriptionByID } from '../store/subscriptions';
import {
  NotFoundError,
  SubscriptionChange,
  SubscriptionDelta,
} from '../store/types';
import { recordUnsubscribeEvent } from '../store/unsubs';

function htmlResponse(html: any, status = 200) {
  return new Response(String(html), {
    status,
    headers: { 'content-type': 'text/html; charset=utf-8' },
  });
}

type ManageCtx = {
  token: string;
  send: Awaited<ReturnType<typeof getSend>>;
  contact: Awaited<ReturnType<typeof getContactByID>>;
  list: Awaited<ReturnType<typeof getListByID>>;
  company: Awaited<ReturnType<typeof getCompanyByID>>;
};

async function loadContext(env: Env, tokenStr: string): Promise<ManageCtx> {
  let t;
  try {
    t = await verify(env.MINIGUN_HMAC_SECRET, tokenStr);
  } catch (err) {
    if (err instanceof InvalidTokenError) throw new Error('Invalid or expired manage link.');
    throw err;
  }
  const snd = await getSend(env.DB, t.sendID).catch(() => {
    throw new Error('Send not found.');
  });
  const sub = await getSubscriptionByID(env.DB, t.subscriptionID).catch(() => {
    throw new Error('Subscription not found.');
  });
  const contact = await getContactByID(env.DB, sub.contact_id).catch(() => {
    throw new Error('Contact not found.');
  });
  const list = await getListByID(env.DB, sub.list_id).catch(() => {
    throw new Error('List not found.');
  });
  if (!list.company_id) {
    throw new Error('This list is not associated with a company; manage page is not available.');
  }
  const company = await getCompanyByID(env.DB, list.company_id).catch(() => {
    throw new Error('Company not found.');
  });
  return { token: tokenStr, send: snd, contact, list, company };
}

export function mountManage(app: Hono<{ Bindings: Env }>) {
  app.get('/manage/:token', async (c) => {
    const tokenStr = c.req.param('token');
    try {
      const ctx = await loadContext(c.env, tokenStr);
      const lists = await getCompanyListsWithSubscription(c.env.DB, ctx.company.id, ctx.contact.id);
      return htmlResponse(
        ManagePage({
          email: ctx.contact.email,
          companyName: ctx.company.name,
          token: tokenStr,
          lists,
        }),
      );
    } catch (err) {
      return htmlResponse(
        ManagePage({
          email: '',
          companyName: '',
          token: tokenStr,
          lists: [],
          error: (err as Error).message,
        }),
        400,
      );
    }
  });

  app.post('/manage/:token', async (c) => {
    const tokenStr = c.req.param('token');
    let ctx: ManageCtx;
    try {
      ctx = await loadContext(c.env, tokenStr);
    } catch (err) {
      return htmlResponse(
        ManagePage({
          email: '',
          companyName: '',
          token: tokenStr,
          lists: [],
          error: (err as Error).message,
        }),
        400,
      );
    }

    const form = (await c.req.parseBody({ all: true }).catch(() => ({}))) as Record<
      string,
      string | File | (string | File)[]
    >;
    const raw = form['list'];
    const checked = new Set<string>(
      Array.isArray(raw) ? raw.map(String) : typeof raw === 'string' ? [raw] : [],
    );

    let companyLists;
    try {
      companyLists = await listsForCompany(c.env.DB, ctx.company.id);
    } catch {
      return htmlResponse(
        ManagePage({
          email: ctx.contact.email,
          companyName: ctx.company.name,
          token: tokenStr,
          lists: [],
          error: 'Failed to load preferences.',
        }),
        500,
      );
    }
    const desired: SubscriptionChange[] = companyLists.map((l) => ({
      list_id: l.id,
      subscribed: checked.has(l.id),
    }));

    let deltas: SubscriptionDelta[];
    try {
      deltas = await applySubscriptionChanges(c.env.DB, ctx.contact.id, desired);
    } catch (err) {
      console.error('apply changes', err);
      return htmlResponse(
        ManagePage({
          email: ctx.contact.email,
          companyName: ctx.company.name,
          token: tokenStr,
          lists: [],
          error: 'Failed to save preferences.',
        }),
        500,
      );
    }

    for (const d of deltas) {
      if (d.was_subbed && !d.now_subbed) {
        try {
          const sub = await getSubscription(c.env.DB, d.list_id, ctx.contact.id);
          await recordUnsubscribeEvent(c.env.DB, ctx.send.id, sub, ctx.contact.email);
        } catch (err) {
          console.error('record unsub event from /manage', err);
        }
      }
    }

    return htmlResponse(
      ManagePage({
        email: ctx.contact.email,
        companyName: ctx.company.name,
        token: tokenStr,
        lists: [],
        done: true,
        deltas,
      }),
    );
  });
}
